packages/create-cloudflare/src/helpers/interactive.ts

```diff
@@ -3,219 +3,260 @@ import { isCancel } from "@clack/prompts";
+import type { C3Arg, C3Args } from "types";
 
 const grayBar = gray(shapes.bar);
 const blCorner = gray(shapes.corners.bl);
 const leftT = gray(shapes.leftT);
 
-export type TextOptions = {
-	renderSubmitted: (value: string) => string;
+export type Option = {
+	label: string;
+	value: string;
+};
+
+export type BasePromptConfig = {
+	// Displayed to the user while prompting for this input
 	question: string;
-	defaultValue: string;
-	acceptDefault: boolean | undefined; // must be specified, but can be undefined (≈ false)
+	// Further clarifies the question
 	helpText?: string;
-	format?: (value: string) => string;
-	validate?: (value: string) => string | void;
+	// The value to use by default
+	defaultValue?: string | boolean;
+	// The status label to be shown after submitting
+	label: string;
+	// Pretty-prints the value in the interactive prompt
+	format?: (value: C3Arg) => string;
+	// Returns a user displayed error if the value is invalid
+	validate?: (value: C3Arg) => string | void;
 };
 
-export const textInput = async (opts: TextOptions) => {
-	const { renderSubmitted, question, defaultValue, validate, acceptDefault } =
-		opts;
-	const helpText = opts.helpText || ``;
-	const format = opts.format || ((val: string) => val);
-
-	const prompt = new TextPrompt({
-		defaultValue: defaultValue,
-		validate,
-		render() {
-			let body = "";
-			switch (this.state) {
-				case "initial":
-					body += `${blCorner} ${bold(question)} ${dim(helpText)}\n`;
-					body += `${space(2)}${gray(format(defaultValue))}\n`;
-					break;
-				case "active":
-					body += `${blCorner} ${bold(question)} ${dim(helpText)}\n`;
-					body += `${space(2)}${format(this.value || dim(defaultValue))}\n`;
-					break;
-				case "submit":
-					body += `${leftT} ${question}\n`;
-					body += `${grayBar} ${renderSubmitted(
-						format(this.value)
-					)}\n${grayBar}`;
-					break;
-				case "error":
-					body += `${leftT} ${status.error} ${dim(this.error)}\n`;
-					body += `${grayBar}\n`;
-					body += `${blCorner} ${question} ${dim(helpText)}\n`;
-					body += `${space(2)}${format(this.value)}\n`;
-					break;
-				default:
-					break;
-			}
-
-			return body;
-		},
-	});
-
-	let value: string;
-	if (acceptDefault) {
-		logRaw(`${leftT} ${question}`);
-		logRaw(`${grayBar} ${renderSubmitted(defaultValue)}\n${grayBar}`);
-		value = defaultValue;
-		validate?.(value);
-	} else {
-		value = (await prompt.prompt()) as string;
-
-		if (isCancel(value)) {
-			cancel("Operation cancelled.");
-			process.exit(0);
-		}
+export type TextPromptConfig = BasePromptConfig & {
+	type: "text";
+};
+export type SelectPromptConfig =
+	| BasePromptConfig & {
+			type: "select";
+			options: Option[];
+	  };
+export type ConfirmPromptConfig =
+	| BasePromptConfig & {
+			type: "confirm";
+			activeText?: string;
+			inactiveText?: string;
+	  };
+
+export type PromptConfig =
+	| TextPromptConfig
+	| ConfirmPromptConfig
+	| SelectPromptConfig;
+
+export const processArgument = async <T>(
```

Comment:
The typings here are really hard to get just right. Hence the `value as T` below.
I think they could be tighter but with the current types, I can't quite work out how.